src/net_processing.cpp

```diff
@@ -3563,32 +3563,48 @@ void ProcessMessage(
     return;
 }
 
+/** Maybe disconnect a peer and discourage future connections from its address.
+ *
+ * @param[in]   pnode           The node to check.
+ * @return                      True if the peer will be disconnected.
+ */
 bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
 {
-    AssertLockHeld(cs_main);
-    CNodeState &state = *State(pnode.GetId());
+    {
+        LOCK(cs_main);
+        CNodeState &state = *State(pnode.GetId());
+
+        // There's nothing to do if the m_should_discourage flag isn't set
+        if (!state.m_should_discourage) return false;
 
-    if (state.m_should_discourage) {
+        // Reset m_should_discourage
         state.m_should_discourage = false;
-        if (pnode.HasPermission(PF_NOBAN)) {
-            LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
-        } else if (pnode.m_manual_connection) {
-            LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
-        } else if (pnode.addr.IsLocal()) {
-            // Disconnect but don't discourage this local node
-            LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", pnode.addr.ToString());
-            pnode.fDisconnect = true;
-        } else {
-            // Disconnect and discourage all nodes sharing the address
-            LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
-            if (m_banman) {
-                m_banman->Discourage(pnode.addr);
-            }
-            connman->DisconnectNode(pnode.addr);
-        }
+    } // cs_main
+
+    if (pnode.HasPermission(PF_NOBAN)) {
+        // Peer has the NOBAN permission flag - log but don't disconnect
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        return false;
+    }
+
+    if (pnode.m_manual_connection) {
+        // Peer is a manual connection - log but don't disconnect
+        LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
+        return false;
+    }
+
+    if (pnode.addr.IsLocal()) {
+        // Peer is on a local address. Disconnect this peer, but don't discourage the local address
+        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", pnode.addr.ToString());
+        pnode.fDisconnect = true;
         return true;
     }
-    return false;
+
+    // Normal case: Disconnect the peer and discourage all nodes sharing the address
+    LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
```

Comment:
f54af5e nit suggestion while refactoring this function

```diff
@@ -3581,27 +3581,29 @@ bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
         state.m_should_discourage = false;
     } // cs_main
 
+    std::string peer{pnode.addr.ToString()};

-        LogPrintf("Warning: not punishing whitelisted peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: not punishing whitelisted peer %s!\n", peer);
 
-        LogPrintf("Warning: not punishing manually-connected peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: not punishing manually-connected peer %s!\n", peer);
 
local address
-        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", pnode.addr.ToString());
+        LogPrintf("Warning: disconnecting but not discouraging local peer %s!\n", peer);
 
-    LogPrintf("Disconnecting and discouraging peer %s!\n", pnode.addr.ToString());
+    LogPrintf("Disconnecting and discouraging peer %s!\n", peer);
```

Reply:
Done (but called the variable `peer_name` since a future commit adds a `Peer` object to this function.